### PR TITLE
CI 2.0 - unified CI for C++, Go and RUST - PHASE1

### PR DESCRIPTION
**Pull request by aviadingo**
## Describe the changes

New CI combines tests for C++/GO/RUST and Examples on a single action.
Tests are divided into curves, fields and hash. each of these can run simultaniously.

Uses a precompiled version of the backend if available on the machine based on the ref to the backend, otherwise recompiles it.

### CODE REVIEW

File: wrappers/rust/icicle-curves/icicle-bls12-377/build.rs [hunk 1]
```diff
@@ -59,46 +59,46 @@ fn main() {
...
 
-    if cfg!(feature = "bw6-761") {
-        // Base config
-        let mut config_bw = Config::new(format!("{}", icicle_src_dir.display()));
-        config_bw
-            .define("CURVE", "bw6_761")
-            .define("FIELD", "bw6_761")
-            .define("HASH", "OFF")
-            .define("CMAKE_INSTALL_PREFIX", &icicle_install_dir);
+    // if cfg!(feature = "bw6-761") {
```
Comment (mickeyasa) [inline wrappers/rust/icicle-curves/icicle-bls12-377/build.rs:62]: why is that commented out?
We still support 761

File: wrappers/rust/icicle-curves/icicle-bw6-761/build.rs [hunk 2]
```diff
@@ -1,23 +1,91 @@
-use std::env;
-use std::path::PathBuf;
+// use std::env;
```
Comment (mickeyasa) [inline wrappers/rust/icicle-curves/icicle-bw6-761/build.rs:1]: please remove commented code

File: wrappers/rust/icicle-curves/icicle-bw6-761/build.rs [hunk 3]
```diff
@@ -1,23 +1,91 @@
...
+        .build_target("install")
+        .build();
+
+    println!("cargo:rustc-link-search={}/lib", icicle_install_dir.display());
+    println!("cargo:rustc-link-lib=icicle_field_bw6_761");
+    println!("cargo:rustc-link-lib=icicle_curve_bw6_761");
+    println!("cargo:rustc-link-lib=icicle_hash");
+    println!("cargo:rustc-link-arg=-Wl,-rpath,{}/lib", icicle_install_dir.display()); // Add RPATH linker arguments
+
+    // default backends dir
```
Comment (mickeyasa) [inline wrappers/rust/icicle-curves/icicle-bw6-761/build.rs:79]: remove line

File: wrappers/rust/icicle-curves/icicle-bw6-761/build.rs [hunk 4]
```diff
@@ -1,23 +1,91 @@
-use std::env;
-use std::path::PathBuf;
+// use std::env;
+// use std::path::PathBuf;
 
-fn main() {
-    // Retrieve environment variables
-    let out_dir = env::var("OUT_DIR").expect("OUT_DIR is not set");
+// fn main() {
```
Comment (mickeyasa) [inline wrappers/rust/icicle-curves/icicle-bw6-761/build.rs:4]: what about the rest of the curves/fields?

**Review by mickeyasa — APPROVED** [review]
great

File: .github/workflows/cpp-golang-rust.yml [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+          - name: bn254
+            build_args: -DG2=ON -DECNTT=ON
+          - name: bls12_381
+            build_args: -DG2=ON -DECNTT=ON
+          - name: bls12_377
+            build_args: -DG2=ON -DECNTT=ON
+          - name: bw6_761
+            build_args: -DG2=ON -DECNTT=ON
+          - name: grumpkin
+            build_args:
```
Comment (LeonHibnik) [inline .github/workflows/cpp-golang-rust.yml:51]: minor: all features are ON by default so maybe it's better to specify explicitly such -DG2=OFF -DECNTT=OFF

**Review by LeonHibnik — APPROVED** [review]
lgtm, left one commet

**Pull request merged**